### ISSUE TITLE
Not working on Windows.

### ISSUE THREAD

**Issue by JRAndreassen** [jjakobh/ssdp-client#3]
Hi...

I'm trying to test the ssdp client on Windows...
I can't see any packet being sent / received in WireShark...
Only Google Chrome noise...
{M-SEARCH * HTTP/1.1
HOST: 239.255.255.250:1900
MAN: "ssdp:discover"
MX: 1
ST: urn:dial-multiscreen-org:service:dial:1
USER-AGENT: Google Chrome/80.0.3987.122 Windows
}
What platform are you targeting

Thanks
  JR

**Comment by jakobhellermann** [comment]
In theory I'm targeting every platform supported by tokio, but I can't test on a windows machine. Can you share your code which isn't working?

**Comment by JRAndreassen** [comment]
I'ts been a while since I tried..
But, I'll check...

**Comment by JRAndreassen** [comment]
Just tried it...
Still nothing:
```rust
use futures::prelude::*;
use std::time::Duration;
use ssdp_client::SearchTarget;

#[async_attributes::main]
async fn main() -> Result<(), ssdp_client::Error> {
    let search_target = SearchTarget::RootDevice;
    let responses = ssdp_client::search(&search_target, Duration::from_secs(3), 2).await?;
    pin_utils::pin_mut!(responses);

    while let Some(response) = responses.next().await {
        println!("{:?}", response?);
    }
    Ok(())
}
```

**Comment by jakobhellermann** [comment]
What version of ssdp-client are you using? I think I switched from async-std to tokio at some point, so running with async_attributes may not work.
Can you try cloning the repo and running the `search`-example?

**Comment by JRAndreassen** [comment]
[package]
name = "ssdp-client"
version = "0.5.3"
authors = ["Jakob Hellermann <jakob.hellermann@protonmail.com>"]
description = "An asynchronous library for discovering, notifying and subscribing to devices and services on a network."
categories = ["network-programming", "asynchronous"]
keywords = ["async", "ssdp", "networking", "protocols"]
repository = "https://github.com/jjakobh/ssdp-client"
documentation = "https://docs.rs/ssdp-client"
readme = "README.md"
license = "MIT/Apache-2.0"
edition = "2018"

27 Apr 2020 07:49:59 -05:00
deb276558d8f8395bfe353ad6394643aa5857102

**Comment by jakobhellermann** [comment]
try replacing `async_attributes::main` with `tokio::main`, I think that should work

**Comment by JRAndreassen** [comment]
Running: "examples/search": Still nothing..
Frame 1: 216 bytes on wire (1728 bits), 216 bytes captured (1728 bits) on interface \Device\NPF_{5F5DACF0-491C-41A2-9CBC-44AC0BBA5E08}, id 0
Ethernet II, Src: IntelCor_24:1a:17 (50:eb:71:24:1a:17), Dst: IPv4mcast_7f:ff:fa (01:00:5e:7f:ff:fa)
Internet Protocol Version 4, Src: 192.168.0.160, Dst: 239.255.255.250
User Datagram Protocol, Src Port: 51649, Dst Port: 1900
Simple Service Discovery Protocol
    M-SEARCH * HTTP/1.1\r\n
        [Expert Info (Chat/Sequence): M-SEARCH * HTTP/1.1\r\n]
        Request Method: M-SEARCH
        Request URI: *
        Request Version: HTTP/1.1
    HOST: 239.255.255.250:1900\r\n
    MAN: "ssdp:discover"\r\n
    MX: 1\r\n
    ST: urn:dial-multiscreen-org:service:dial:1\r\n
    USER-AGENT: Google Chrome/83.0.4103.106 Windows\r\n
    \r\n
    [Full request URI: http://239.255.255.250:1900*]
    [HTTP request 1/2]
    [Next request in frame: 2]

I'll have to look into it deeper when I have time..
I'll get back to you.
Thanks

**Comment by JRAndreassen** [comment]
Sorry, copied the wrong one from Wireshark...
There are some low-level networking routines on Windows that are not working / implemented..
I had the same problem trying to get ICMP packets going ...
Sorry, I'm deep in debugging another problem at the moment...

**Comment by jakobhellermann** [comment]
no problem

**Comment by dheijl** [comment]
I'm having a similar problem on Windows 10 pro: discovery is not working, no broadcasts on the network according to Wireshark.
But ... usually after a couple of hours suddenly it starts working and from then on it continues to work until a reboot of Windows. 
I'm using rupnp by the way, not the ssdp client directly.

**Comment by dheijl** [comment]
The reason seems to be that Windows 10 is multihomed so that the address that is used for the broadcast send is not guaranteed to be your local ip address, it can be any of the virtual interfaces instead. If you replace the bind to 0.0.0.0. (INADDR_ANY) in search() with a bind to the local ethernet ip address it works fine.

**Comment by dheijl** [comment]
I made the following change in search.rs so that it now always works in Windows:

```rust
    //let bind_addr: SocketAddr = ([0, 0, 0, 0], 0).into(); // this line replaced by the following code
    // get local ip address (borrowed from https://docs.rs/local_ipaddress/0.1.3/local_ipaddress/)
    let any: SocketAddr = ([0, 0, 0, 0], 0).into();
    let socket = UdpSocket::bind(any).await.expect("Could not bind the UDP socket to INADDR_ANY");
    let googledns : SocketAddr = ([8, 8, 8, 8], 80).into();
    socket.connect(googledns).await.expect("No network connectivity");
    let bind_addr = socket.local_addr().expect("Could not obtain local ip address for udp broadcast socket");

```

With this change the code still works on Linux (Ubuntu 20.04).

**Comment by jakobhellermann** [comment]
@dheijl what is the point of connecting to the google DNS server?
I just tried your snippet without it and it worked fine.

**Comment by dheijl** [comment]
It doesn't always work on Windows 10 if you have virtualization enabled because you don't know on what interface the broadcast will be sent (a virtual interface can be selected). So I make sure that the interface that is connected to the real network is used.

**Comment by jakobhellermann** [comment]
This is now fixed in 0.6.2, thanks dheijl for figuring out the solution.
